src/Roslyn.Suppressions/Roslyn.Suppressions.UnitTests/SuppressionCheckerTests.cs

```diff
@@ -0,0 +1,403 @@
+        /// </summary>
+        /// <remarks>
+        /// The issue will match the diagnostic, unless the file path is overridden.
+        /// The diagnostic must have a valid location and source tree.
+        /// </remarks>
+        private static SonarQubeIssue CreateIssueFromDiagnostic(Diagnostic diagnostic,
+            string overrideFilePath = null)
+        {
+            // The issue will match because the line number is the same (so the hash is irrelevant)
+            var sonarLine = diagnostic.Location.GetLineSpan().EndLinePosition.Line + 1;
+
+            var issuePath = overrideFilePath ?? diagnostic.Location.SourceTree.FilePath;
+            var issueHash = System.Guid.NewGuid().ToString();
+
+            return CreateIssue("any", diagnostic.Id, issuePath, sonarLine, issueHash);
+        }
+
+        private static SonarQubeIssue CreateIssueWithTextRange(string ruleId,
+            string path,
+            string hash,
+            IssueTextRange textRange)
+        {
+            return new SonarQubeIssue("any issue key",
+                path,
+                hash,
+                "message",
+                "moduleKey",
+                ruleId,
+                true,
+                SonarQubeIssueSeverity.Blocker,
+                DateTimeOffset.Now,
+                DateTimeOffset.Now,
+                textRange,
+                null);
+        }
+
+        #region Diagnostic helper methods
+
+        private static Location CreateNonSourceLocation()
+        {
+            var nonSourceLocation = Location.Create("dummyFilePath.cs", new TextSpan(), new LinePositionSpan());
+            nonSourceLocation.IsInSource.Should().BeFalse();
+            nonSourceLocation.Kind.Should().NotBe(LocationKind.SourceFile);
+            return nonSourceLocation;
+        }
+
+        /// <summary>
+        /// Returns a Location that is in a source file, backed by a valid syntax tree.
+        /// </summary>
+        /// <param name="fileText">The contents of the entire file (optional)</param>
+        /// <param name="selectedText">The specific text in the file the location points to (optional)</param>
+        private static Location CreateSourceFileLocation(string filePath,
+            string fileText = "any", string selectedText = "any")
+        {
+            // Mocking source location so that it appears to be backed by a real syntax tree, with
+            // the required "GetLineSpan()", "GetText().Lines" etc methods working correctly is hard.
+
+            // The simplest thing to do is create a real C# syntax tree. It doesn't matter if the
+            // contents of the fake file are not valid C# - the parser will still produce a valid
+            // document.
+
+            var syntaxTree = Microsoft.CodeAnalysis.CSharp.CSharpSyntaxTree.ParseText(fileText, path: filePath);
+            var selectedSpan = CreateSpan(fileText, selectedText);
+            var location = Location.Create(syntaxTree, selectedSpan);
```

Comment:
I spent ages trying to mock the `Location` and dependent objects correctly, then realised it was simpler just to ask Roslyn to produce real objects for us.